core/src/main/java/com/datastrato/gravitino/authorization/PermissionManager.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.authorization;
+
+import static com.datastrato.gravitino.authorization.AuthorizationUtils.GROUP_DOES_NOT_EXIST_MSG;
+import static com.datastrato.gravitino.authorization.AuthorizationUtils.USER_DOES_NOT_EXIST_MSG;
+
+import com.datastrato.gravitino.Entity;
+import com.datastrato.gravitino.EntityStore;
+import com.datastrato.gravitino.exceptions.NoSuchEntityException;
+import com.datastrato.gravitino.exceptions.NoSuchGroupException;
+import com.datastrato.gravitino.exceptions.NoSuchRoleException;
+import com.datastrato.gravitino.exceptions.NoSuchUserException;
+import com.datastrato.gravitino.exceptions.RoleAlreadyExistsException;
+import com.datastrato.gravitino.meta.AuditInfo;
+import com.datastrato.gravitino.meta.GroupEntity;
+import com.datastrato.gravitino.meta.RoleEntity;
+import com.datastrato.gravitino.meta.UserEntity;
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import com.google.common.collect.Lists;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * PermissionManager is used for managing the logic the granting and revoking roles. Role is used
+ * for manging permissions. GrantManager will filter the invalid roles, too.
+ */
+class PermissionManager {
+  private static final Logger LOG = LoggerFactory.getLogger(PermissionManager.class);
+
+  private final EntityStore store;
+
+  PermissionManager(EntityStore store) {
+    this.store = store;
+  }
+
+  public boolean grantRoleToUser(String metalake, String role, String user) {
+    try {
+      AuthorizationUtils.checkMetalakeExists(metalake);
+
+      checkRoleExists(metalake, role);
+
+      RoleEntity roleEntity =
+          AuthorizationUtils.getRoleEntity(NameIdentifierUtils.ofRole(metalake, role));
+
+      store.update(
+          NameIdentifierUtils.ofUser(metalake, user),
+          UserEntity.class,
+          Entity.EntityType.USER,
+          userEntity -> {
+            List<RoleEntity> roleEntities =
+                AuthorizationUtils.getValidRoles(
+                    metalake, userEntity.roles(), userEntity.roleIds());
```

Comment:
Change roles to `roleNames` for better understanding.

Reply:
Done.